### PR TITLE
Mark Dependency.optional as deprecated.

### PR DESCRIPTION
**Pull request by vkryachko**

### CODE REVIEW

File: firebase-components/src/main/java/com/google/firebase/components/Dependency.java [hunk 1]
```diff
@@ -47,30 +47,82 @@ private Dependency(Class<?> anInterface, @Type int type, @Injection int injectio
...
+  @Deprecated
   public static Dependency optional(Class<?> anInterface) {
     return new Dependency(anInterface, Type.OPTIONAL, Injection.DIRECT);
   }
 
+  /**
+   * Declares a deferred dependency.
+   *
+   * <p>Such dependencies are optional and may not be present be default. But they can become
+   * available if a dynamic module is installed that they are contained in.
```
Comment (rlazo) [inline firebase-components/src/main/java/com/google/firebase/components/Dependency.java:68]:
```suggestion
   * available if a dynamic module that contains them is installed.
```

File: firebase-components/src/main/java/com/google/firebase/components/Dependency.java [hunk 2]
```diff
@@ -47,30 +47,82 @@ private Dependency(Class<?> anInterface, @Type int type, @Injection int injectio
...
+   */
+  @Deprecated
   public static Dependency optional(Class<?> anInterface) {
     return new Dependency(anInterface, Type.OPTIONAL, Injection.DIRECT);
   }
 
+  /**
+   * Declares a deferred dependency.
+   *
+   * <p>Such dependencies are optional and may not be present be default. But they can become
```
Comment (rlazo) [inline firebase-components/src/main/java/com/google/firebase/components/Dependency.java:67]:
```suggestion
   * <p>Such dependencies are optional and may not be present by default. But they can become
```

File: firebase-components/src/main/java/com/google/firebase/components/Dependency.java [hunk 3]
```diff
@@ -47,30 +47,82 @@ private Dependency(Class<?> anInterface, @Type int type, @Injection int injectio
...
   public static Dependency setOf(Class<?> anInterface) {
     return new Dependency(anInterface, Type.SET, Injection.DIRECT);
   }
 
+  /**
+   * Declares an optional dependency.
+   *
+   * <p>Optional dependencies can be missing at runtime(being {@code null}) and dependents must be
+   * ready to handle that.
+   */
```
Comment (rlazo) [inline firebase-components/src/main/java/com/google/firebase/components/Dependency.java:102]: This isn't deprecated?

Comment (vkryachko) [reply firebase-components/src/main/java/com/google/firebase/components/Dependency.java:102]: Nope, this is now the recommended optional dependency, see the deprecation message in `Dependency.optional()`.

**Review by rlazo — APPROVED** [review]

**Pull request merged**